ZOOKEEPER-1105: wait for server response in C client zookeeper_close

symat opened the pull request:
**Thanks for Lincoln Lee for the original fix!**

In the current implementation, we always get a WARN in server side
("EndOfStreamException: Unable to read additional data from client")
whenever we close a zookeeper handler from the C client. This also happens
in the end of every execution of the command line C client.

The reason is that currently we don't wait for the response from the server
when we initiate the closing of the client connection, and we terminate
the socket on the client side too early.

I tested the patch both on linux and windows. I also tested it both with
NIO and Netty server side socket implementations.

anmolnar commented on zookeeper-client/zookeeper-client-c/src/zookeeper.c:
```diff
@@ -3495,6 +3495,51 @@ static int add_multi_completion(zhandle_t *zh, int xid, void_completion_t dc,
+    } else if (ret < 0) {
+        LOG_WARN(LOGCALLBACK(zh), "System error (%d) happened while waiting for server's reply, sessionId=%#llx\n",
+            ret, zh->client_id.client_id);
+    } else {
```
I suspect that this is the positive branch. Why not removing it entirely?

symat replied:
you are right, this is the positive case (when the server responded to our 'close' request), when we don't need to do anything. I removed the empty block.

suhasdantkale commented on zookeeper-client/zookeeper-client-c/src/zookeeper.c:
```diff
@@ -3495,6 +3495,49 @@ static int add_multi_completion(zhandle_t *zh, int xid, void_completion_t dc,
+#else
+    FD_ZERO(&rfds);
+    FD_SET(zh->fd->sock , &rfds);
+    ret = select(0, &rfds, NULL, NULL, &waittime);
```
Minor - Instead of select on 0, could we make it "zh->fd->sock + 1"?

symat commented:
@anmolnar @nkalmar what do you think? this is a minor cleanup thing, but it fixes an issue open for 12 years now :) And also helps supportability potentially, by getting the rid of unnecessary warnings in the server logs.
Should we ask anyone else for a review?

suhasdantkale commented on zookeeper-client/zookeeper-client-c/src/zookeeper.c:
```diff
@@ -3495,6 +3495,51 @@ static int add_multi_completion(zhandle_t *zh, int xid, void_completion_t dc,
+    struct pollfd fd_s[1];
+#else
+    fd_set rfds;
+    struct timeval waittime = {timeout_ms / 1000, (timeout_ms % 1000) * 1000000};
```
@symat ,
should it be ?:
struct timeval waittime = {timeout_ms / 1000, (timeout_ms % 1000) * 1000};
instead of
struct timeval waittime = {timeout_ms / 1000, (timeout_ms % 1000) * 1000000};

symat replied:
You are right, nice catch, thanks! :)
timeval indeed expects microsec, not nanosec.
I will fix in the next commit.

suhasdantkale commented on zookeeper-client/zookeeper-client-c/src/zookeeper.c:
```diff
@@ -3495,6 +3495,51 @@ static int add_multi_completion(zhandle_t *zh, int xid, void_completion_t dc,
+        LOG_WARN(LOGCALLBACK(zh), "System error (%d) happened while waiting for server's reply, sessionId=%#llx\n",
+            ret, zh->client_id.client_id);
+    } else {
+        // do nothing
```
Should we read the data that server sent us before closing?
I will check what does serve send as response to ZOO_CLOSE_OP(i.e OpCode.closeSession from server side).

symat replied:
In general I am not sure if it matters at this stage. I mean we will not behave differently regardless what the server responds. (I think we will actually get back an ACK in https://github.com/apache/zookeeper/blob/e4758ba70b35dd38ff0bba9e534790391f53b050/zookeeper-server/src/main/java/org/apache/zookeeper/server/FinalRequestProcessor.java#L330 ). I can verify it by wireshark if needed. Do you think it matters if we just skip the processing of the response?

suhasdantkale replied:
You are right Mate. It doesn't matter in the context of this bug. I was just trying to understand. You can please go ahead with your changes.

suhasdantkale commented:
Hi @symat ,
Overall, the changes look good to me
I've left some minor comments/questions.

symat commented:
@suhasdantkale thanks for the comments!
I fixed the first bug you found, and also rebased to the current master branch as the C-client SASL support got merged recently. I tested the patch manually with the C CLI agaist a running ZooKeeper 3.5.5 and it still works (no warnings in the server logs).

suhasdantkale commented:
Thanks Mate.
LGTM.

nkalmar commented:
Merged to master, 3.6 and 3.5 branch. Thanks @symat and Lincoln!
